internal/schemautil/service.go

```diff
@@ -1113,3 +1115,38 @@ func ContainsRedactedCreds(config map[string]any) error {
 	}
 	return nil
 }
+
+var (
+	servicePoweredMap    sync.Map
+	servicePoweredMutex  sync.Mutex
+	errServicePoweredOff = fmt.Errorf("the service is powered off")
+)
+
+// CheckServiceIsPowered checks if the service is powered on.
+// When a service is powered off, some endpoints might return "503: An error occurred. Please try again later."
+func CheckServiceIsPowered(ctx context.Context, client avngen.Client, project, serviceName string) error {
+	serviceKey := filepath.Join(project, serviceName)
+	value, ok := servicePoweredMap.Load(serviceKey)
+	if ok {
+		if !value.(bool) {
+			return errServicePoweredOff
+		}
+		return nil
+	}
+
+	servicePoweredMutex.Lock()
+	defer servicePoweredMutex.Unlock()
+
+	s, err := client.ServiceGet(ctx, project, serviceName)
+	if err != nil {
+		return err
+	}
+
+	powered := s.State != service.ServiceStateTypePoweroff
+	servicePoweredMap.Store(serviceKey, powered)
+	if powered {
+		return nil
+	}
+
+	return CheckServiceIsPowered(ctx, client, project, serviceName)
```

Comment:
I see my mistake. Thats true, there are exit from recursion, but still don't get the necessity of it.

Reply:
The recursion would return `ErrServicePoweredOff` if the stored value is not true.
Either I put this check before and after the `ServiceGet` or I call the function recursively to keep things DRY.